### PR TITLE
[Multiplatform] Migrate `ProjectDescription.Target` to support multiple platforms

### PR DESCRIPTION
**Pull request by waltflanagan**
Very first exposure of multiplatform work in ProjectDescription.  All feedback on API design welcome.  

Addresses #5515 

### Short description 📝

Complete adding support for multiple platform targets in Tuist by exposing the API in `ProjectDescription.Target`

In order to minimize disruption and maintain source compatibility, we will deprecate the single platform `init` in preference of `init(destinations:deploymentTargets:)`.  We expose an analog of `Platform` and `DeploymentTarget` for source compatibility but they should no longer be relied upon.

To support multiplatform targets, we also migrate SPM dependencies via `Dependencies.swift` to use multiplatform targets instead of the single platform `Framework_ios` patterns.

When generating the resulting `xcodeproj`, platform filters are applied to `PBXBuildFile` instances via `LinkGenerator`

### How to test the changes locally 🧐

Run tests. All should pass.

### Contributor checklist ✅

- [X] The code has been linted using run `make workspace/lint-fix`
- [X] The change is tested via unit testing or acceptance testing, or both
- [X] The title of the PR is formulated in a way that is usable as a changelog entry
- [x] In case the PR introduces changes that affect users, the documentation has been updated

### Reviewer checklist ✅

- [ ] The code architecture and patterns are consistent with the rest of the codebase
- [ ] Reviewer has checked that, if needed, the PR includes the label `changelog:added`, `changelog:fixed`, or `changelog:changed`, and the title is usable as a changelog entry

### CODE REVIEW

File: Sources/ProjectDescription/TargetDependency.swift [hunk 1]
```diff
@@ -47,29 +47,29 @@ public enum TargetDependency: Codable, Hashable {
     ///
     /// - Parameters:
     ///   - name: Name of the target to depend on
-    case target(name: String)
+    case target(name: String, platformFilters: PlatformFilters = .all)
```
Comment (danyf90) [inline Sources/ProjectDescription/TargetDependency.swift:50]: I think it's strange if all cases require the same associated values, which I also expect to be relatively rarely used 🤔 

Not 100% sure, but maybe we can do something like:
```
Target(
  …,
  dependencies: [
    .target('A'),
    .target('B').only(on: .iOS),
  ],
)
```
or something similar?

Comment (waltflanagan) [reply Sources/ProjectDescription/TargetDependency.swift:50]: what concrete type would that resolve to within that array? these being defined as `enum` types constrain our design space a bit.  i think we could add some instance methods to these to provide some syntactic sugar like that but in the end given we're maintaining source compatibility, I wasnt able to see any other ways to achieve what we need without adding this new associated value across all cases.  

This applies both to the API we expose to users but also how packages are mapped into `ProjectDescription` types during `fetch`

Comment (kwridan) [reply Sources/ProjectDescription/TargetDependency.swift:50]: x-posting some thoughts on platform filters which are also applicable here https://github.com/tuist/tuist/pull/5557#pullrequestreview-1745191008

Comment (kwridan) [reply Sources/ProjectDescription/TargetDependency.swift:50]: > think it's strange if all cases require the same associated values, which I also expect to be relatively rarely used 🤔

👍 anecdotally from our project (a graph with 1000s of edges) only 3 of them have filters specified - which may ago at some point too

File: Sources/TuistCore/Graph/GraphLoader.swift [hunk 2]
```diff
@@ -105,17 +103,20 @@ public final class GraphLoader: GraphLoading {
...
+        let targetDependency = GraphDependency.target(name: name, path: path)
+        let dependencies: [GraphDependency] = try target.dependencies.compactMap { dependency in
+            guard let graphDep = try loadDependency(
                 path: path,
                 forPlatforms: target.supportedPlatforms,
                 dependency: dependency,
                 cache: cache
-            )
+            ) else { return nil }
+            cache.dependencyPlatformFilters[(targetDependency, graphDep)] = dependency.platformFilters
```
Comment (kwridan) [inline Sources/TuistCore/Graph/GraphLoader.swift:114]: Given we don't expect this to be heavily used (there may even be graphs that don't specify any filters at all!) we can only store explicitly specified filters as opposed to all filters for all edges. 

This means we can add some very easy initial optimisations such as skipping the platform filter traversals all together if there are no filters specified at all.

**Review by kwridan — COMMENTED** [review]
thanks @waltflanagan

File: Sources/ProjectDescription/TargetDependency.swift (hunk 1 above)
Comment (kwridan) [reply Sources/ProjectDescription/TargetDependency.swift:50]: >  I wasnt able to see any other ways to achieve what we need without adding this new associated value across all cases.

Agreed the current approach of adding `platformFilters` is the easiest path to Tuist 3 source compatibility

**Review by kwridan — COMMENTED** [review]
Awesome, thanks for the all the updates @waltflanagan 👍 

The introduction of `Condition` and `CombinationResults` makes things a lot more clearer and easier to follow 

I retested the changes against our project and looks like there are no diffs / regressions 🎉 

I believe all my feedback points from https://github.com/tuist/tuist/pull/5557#pullrequestreview-1745191008 have been addressed, thanks again!

**Pull request merged**